SMCP and ROCm 7.0 compatibility

**Bug summary**

ROCm 7.0 beta makes `warpSize` into a [run-time constant](https://rocm.docs.amd.com/en/docs-7.0-beta/preview/release.html#warpsize) instead of a compile-time one. This breaks its use in

https://github.com/AdaptiveCpp/AdaptiveCpp/blob/706300a46fac5056ef2be17f876b8c6c8d1f2727/include/hipSYCL/sycl/libkernel/detail/thread_hierarchy.hpp#L124-L130

**To Reproduce**

Build ACpp `develop` with [ROCm 7.0 Beta](https://rocm.docs.amd.com/en/docs-7.0-beta/preview/install/rocm.html) (does not matter whether using ROCm LLVM or mainline LLVM), try building any SYCL program with `--acpp-targets=hip:gfxABCD`, get

```
/home/user/local/bin/../include/AdaptiveCpp/hipSYCL/glue/hip/../generic/hiplike/hiplike_kernel_launcher.hpp:247:52: error: non-type template argument is not a constant expression
  247 |                          nested_range<static_range<__acpp_warp_size>>>;
      |                                                    ^~~~~~~~~~~~~~~~
/home/user/local/bin/../include/AdaptiveCpp/sycl/../hipSYCL/sycl/libkernel/detail/thread_hierarchy.hpp:129:28: note: expanded from macro '__acpp_warp_size'
  129 |   #define __acpp_warp_size warpSize
      |                            ^~~~~~~~
/home/user/local/bin/../include/AdaptiveCpp/hipSYCL/glue/hip/../generic/hiplike/hiplike_kernel_launcher.hpp:247:52: note: non-constexpr function 'operator int' cannot be used in a constant expression
/home/user/local/bin/../include/AdaptiveCpp/sycl/../hipSYCL/sycl/libkernel/detail/thread_hierarchy.hpp:129:28: note: expanded from macro '__acpp_warp_size'
  129 |   #define __acpp_warp_size warpSize
      |                            ^
/opt/rocm-7.0.0/include/hip/amd_detail/amd_warp_functions.h:89:5: note: declared here
   89 |     operator int() const noexcept {
      |     ^

```

**Expected behavior**

The code compiles.

**Describe your setup**

```
acpp [AdaptiveCpp compilation driver], Copyright (C) 2018-2025 Aksel Alpay and the AdaptiveCpp project
  AdaptiveCpp version: 25.02.0+git.706300a4.20250731.branch.develop.dirty
  Installation root: /home/user/local
  Plugin LLVM version: 19, can accelerate CPU: True
  Available runtime backends:
     librt-backend-cuda.so
     librt-backend-omp.so
     librt-backend-hip.so
     librt-backend-ocl.so


Full configuration [can be overridden using environment variables or command line arguments]:
    version-major: 25
    version-minor: 02
    version-patch: 0
    version-suffix: +git.706300a4.20250731.branch.develop.dirty
    plugin-linked-into-llvm: false
    plugin-llvm-version-major: 19
    plugin-with-cpu-acceleration: true
    plugin-with-sscp-compiler: true
    default-clang: /usr/lib/llvm-19/bin/clang++
    default-targets: generic
    default-cpu-cxx: /usr/bin/clang++-19
    default-rocm-path: /opt/rocm-7.0.0
    default-use-bootstrap-mode: false
    default-is-dryrun: false
    default-use-accelerated-cpu: true
    default-clang-include-path: /usr/lib/llvm-19/lib/clang/19/include/..
    default-sequential-link-line: -L/usr/lib/x86_64-linux-gnu -lboost_context -lboost_fiber -Wl,-rpath=/usr/lib/x86_64-linux-gnu
    default-sequential-cxx-flags: -I/usr/include -D_ENABLE_EXTENDED_ALIGNED_STORAGE
    default-omp-link-line: -L/usr/lib/x86_64-linux-gnu -lboost_context -lboost_fiber -Wl,-rpath=/usr/lib/x86_64-linux-gnu -fopenmp
    default-omp-cxx-flags: -I/usr/include -fopenmp -D_ENABLE_EXTENDED_ALIGNED_STORAGE
    default-is-explicit-multipass: false
    default-save-temps: false
    default-rocm-link-line: -Wl,-rpath=$ACPP_ROCM_PATH/lib -Wl,-rpath=$ACPP_ROCM_PATH/hip/lib -L/opt/rocm-7.0.0/lib -L/opt/rocm-7.0.0/hip/lib -lamdhip64
    default-rocm-cxx-flags: -isystem $ACPP_PATH/include/AdaptiveCpp/hipSYCL/std/hiplike -isystem /usr/lib/llvm-19/lib/clang/19/include/.. -U__FLOAT128__ -U__SIZEOF_FLOAT128__ -I$ACPP_ROCM_PATH/include -I$ACPP_ROCM_PATH/include --rocm-device-lib-path=$ACPP_ROCM_PATH/amdgcn/bitcode --rocm-path=$ACPP_ROCM_PATH -fhip-new-launch-api -mllvm -amdgpu-early-inline-all=true -mllvm -amdgpu-function-calls=false -D__HIP_ROCclr__
    default-nvcxx: (unconfigured)
    default-cuda-path: /usr/local/cuda
    default-cuda-link-line: -Wl,-rpath=$ACPP_CUDA_LIB_PATH -L$ACPP_CUDA_LIB_PATH -lcudart
    default-cuda-cxx-flags: -U__FLOAT128__ -U__SIZEOF_FLOAT128__ -isystem $ACPP_PATH/include/AdaptiveCpp/hipSYCL/std/hiplike
```

**Additional context**

`warpSize` in ROCm 7 is implemented as:

```cpp
inline __device__ const struct final {
  __device__
  __attribute__((always_inline, const))
    operator int() const noexcept {
      return __builtin_amdgcn_wavefrontsize();
    }
} warpSize{};
```

The [official solution for those wanting a compile-time constant](https://rocm.docs.amd.com/projects/HIP/en/docs-develop/how-to/hip_cpp_language_extensions.html#warpsize) is

```cpp
#if defined(__GFX8__) || defined(__GFX9__)
  #define WarpSize 64
#else
  #define WarpSize 32
#endif
```

--- a/include/hipSYCL/sycl/libkernel/detail/thread_hierarchy.hpp
+++ b/include/hipSYCL/sycl/libkernel/detail/thread_hierarchy.hpp
@@ -126,7 +126,11 @@ HIPSYCL_DEFINE_BUILTIN_VARIABLE_QUERY(__acpp_get_ngroups_z,
   // for now
   #define __acpp_warp_size 32
 #else
-  #define __acpp_warp_size warpSize
+  #if defined(__GFX8__) || defined(__GFX9__)
+    #define __acpp_warp_size 64
+  #else
+    #define __acpp_warp_size 32
+  #endif
 #endif
 
 #endif

